src/coord/src/timestamp.rs

```diff
@@ -333,60 +394,117 @@ fn generate_ts_updates_from_debezium(
                 }
             }
         }
+    } else {
+        bail!("Expected record type for value, got {:?}", value);
     }
+
+    Ok(())
 }
 
 /// A debezium record contains a set of update counts for each topic that the transaction
 /// updated. This function extracts the set of (topic, update_count) as a vector if
 /// processing an END message. It returns NONE otherwise.
-fn parse_debezium(record: Vec<(String, Value)>) -> Option<Vec<(String, i64)>> {
+fn parse_debezium(record: Vec<(String, Value)>) -> anyhow::Result<Option<Vec<(String, i64)>>> {
```

Comment:
this function can be made substantially simpler with multi-level matches
```
diff --git a/src/coord/src/timestamp.rs b/src/coord/src/timestamp.rs
index 5efe67fa4..9d2c1d925 100644
--- a/src/coord/src/timestamp.rs
+++ b/src/coord/src/timestamp.rs
@@ -408,50 +408,32 @@ fn parse_debezium(record: Vec<(String, Value)>) -> anyhow::Result<Option<Vec<(St
     let mut result = vec![];
     let mut event_count = 0;
     for (key, value) in record {
-        if key == "status" {
-            if let Value::String(status) = value {
-                match status.as_str() {
-                    "BEGIN" => return Ok(None),
-                    "END" => (),
-                    _ => {
-                        bail!(
-                            "Failed to parse Debezium transaction message. Invalid status '{}'",
-                            status
-                        )
-                    }
+        match (key.as_str(), value) {
+            ("status", Value::String(status)) => match status.as_str() {
+                "BEGIN" => return Ok(None),
+                "END" => continue,
+                _ => {
+                    bail!(
+                        "Failed to parse Debezium transaction message. Invalid status '{}'",
+                        status
+                    )
                 }
-            } else {
-                bail!(
-                        "Failed to parse Debezium transaction message. Expected String for field 'status', got {:?}",
-                        value
-                    );
-            }
-        } else if key == "data_collections" {
-            if let Value::Union { inner: value, .. } = value {
-                if let Value::Array(items) = *value {
+            },
+            ("data_collections", Value::Union { inner, .. }) => {
+                if let Value::Array(items) = *inner {
                     for v in items {
                         if let Value::Record(item) = v {
                             let mut value: Option<String> = None;
                             let mut write_count: Option<i64> = None;
                             for (k, v) in item {
-                                if k == "data_collection" {
-                                    if let Value::String(data) = v {
+                                match (k.as_str(), v) {
+                                    ("data_collection", Value::String(data)) => {
                                         value = Some(data);
-                                    } else {
-                                        bail!(
-                                            "Failed to parse Debezium transaction message. Expected string for 'data_collection', got {:?}",
-                                            v
-                                        );
                                     }
-                                } else if k == "event_count" {
-                                    if let Value::Long(e) = v {
-                                        write_count = Some(e);
-                                    } else {
-                                        bail!(
-                                            "Failed to parse Debezium transaction message. Expected long for 'event_count', got {:?}",
-                                            v
-                                        );
+                                    ("event_count", Value::Long(count)) => {
+                                        write_count = Some(count);
                                     }
+                                    (key, value) => bail!("Failed to parse Debezium transaction message. Invalid key or value in data_collections. Key: {:?} Value:{:?}", key, value),
                                 }
                             }
                             match (value, write_count) {
@@ -472,31 +454,21 @@ fn parse_debezium(record: Vec<(String, Value)>) -> anyhow::Result<Option<Vec<(St
                 } else {
                     bail!(
                         "Failed to parse Debezium transaction message. Array expected, got {:?}",
-                        value
-                    );
+                        inner
+                    )
                 }
-            } else {
-                bail!(
-                    "Failed to parse Debezium transaction message. Union of Null/Array expected, got {:?}",
-                    value
-                );
             }
-        } else if key == "event_count" {
-            if let Value::Union { inner: value, .. } = value {
-                if let Value::Long(count) = *value {
+            ("event_count", Value::Union { inner, .. }) => {
+                if let Value::Long(count) = *inner {
                     event_count = count;
                 } else {
                     bail!(
-                        "Failed to parse Debezium event count. Long expected, got {:?}",
-                        value
-                    );
+                        "Failed to parse Debezium transaction message. Long expected, got {:?}",
+                        inner
+                    )
                 }
-            } else {
-                bail!(
-                    "Failed to parse Debezium event count. Union expected, got {:?}",
-                    value
-                );
             }
+            (k, v) => bail!("Invalid key/val combination key: {:?} value: {:?}", k, v),
         }
     }
     // TODO (chris): why can't we simply assert that event_count is non-zero?
```

also the previous checks within records were non-exhaustive -- there could be a key that is not `event_count` or `data_collection` -- added an exhaustiveness check

Reply:
The multi-level match arms do make a nice difference, thanks! I'll make that change!

The first version of this PR changed less of this code, so I was trying to remain in style with the existing code.

Reply:
Done!